File: src/main/java/net/cjsah/console/command/CommandManager.java
package net.cjsah.console.command;

import net.cjsah.console.Console;
import net.cjsah.console.command.argument.Argument;
import net.cjsah.console.command.builder.LiteralArgumentBuilder;
import net.cjsah.console.command.builder.RequiredArgumentBuilder;
import net.cjsah.console.command.source.CommandSource;
import net.cjsah.console.exceptions.CommandException;

import java.util.function.Consumer;

public class CommandManager {

    private static final Dispatcher DISPATCHER = new Dispatcher();

    public static LiteralArgumentBuilder literal(String literal) {
        return LiteralArgumentBuilder.literal(literal);
    }

    public static <T> RequiredArgumentBuilder<T> argument(String literal, Argument<T> type) {
        return RequiredArgumentBuilder.argument(literal, type);
    }

    public static int execute(String command, CommandSource<?> source) {
        try {
            return DISPATCHER.execute(command, source);
        }catch (CommandException e) {
            if (e.getMessage() != null) Console.INSTANCE.getLogger().error(e.getMessage());
        }catch (Exception e) {
            e.printStackTrace();
        }
        return 0;
    }

    public static void register(Consumer<Dispatcher> command) {
        command.accept(DISPATCHER);
    }

}

File: src/main/kotlin/net/cjsah/console/command/CommandManager.kt
package net.cjsah.console.command

import net.cjsah.console.Logger
import net.cjsah.console.command.argument.Argument
import net.cjsah.console.command.builder.LiteralArgumentBuilder
import net.cjsah.console.command.builder.RequiredArgumentBuilder
import net.cjsah.console.command.source.CommandSource
import net.cjsah.console.plugin.Plugin
import java.util.function.Consumer

object CommandManager {
    @JvmStatic
    private val DISPATCHER = Dispatcher()

    @JvmStatic
    fun literal(literal: String): LiteralArgumentBuilder =
        LiteralArgumentBuilder.literal(literal)


    @JvmStatic
    fun <T> argument(literal: String, type: Argument<T>): RequiredArgumentBuilder<T> =
        RequiredArgumentBuilder.argument(literal, type)


    @JvmStatic
    suspend fun execute(command: String, source: CommandSource<*>): Int {
        try {
            return DISPATCHER.execute(command, source)
        } catch (e: Exception) {
            e.message?.let { Logger.error(it) }
        }
        return 0
    }

    @JvmStatic
    fun register(command: Consumer<Dispatcher>) = command.accept(DISPATCHER)

    @JvmStatic
    internal fun deregister(plugin: Plugin) = DISPATCHER.deregister(plugin)

}